static/app/components/inputGroup.tsx

```diff
@@ -0,0 +1,219 @@
+import {createContext, forwardRef, useContext, useEffect, useRef, useState} from 'react';
+import {useTheme} from '@emotion/react';
+import styled from '@emotion/styled';
+
+import _TextArea from 'sentry/components/forms/controls/textarea';
+import _Input, {InputProps} from 'sentry/components/input';
+import space from 'sentry/styles/space';
+import {FormSize} from 'sentry/utils/theme';
+import useMedia from 'sentry/utils/useMedia';
+
+interface InputContext {
+  /**
+   * Props to be passed to `Input`. When `InputGroup` is used, it is
+   * recommended to add input props (`size`, `disabled`) to `InputGroup`
+   * rather than `Input`, so that other elements in the group
+   * (`InputLeadingItems`, `InputTrailingItems`) also know about them.
+   */
+  inputProps: Partial<InputProps>;
+  /**
+   * Width of the leading items wrap, to be added to `Input`'s padding.
+   */
+  leadingWidth?: number;
+  setLeadingWidth?: (width: number) => void;
+  setTrailingWidth?: (width: number) => void;
+  /**
+   * Width of the trailing items wrap, to be added to `Input`'s padding.
+   */
+  trailingWidth?: number;
+}
+export const InputGroupContext = createContext<InputContext>({inputProps: {}});
+
+/**
+ * Context provider for input group. To be used alongisde `Input`, `InputLeadingItems`,
+ * and `InputTrailingItems`:
+ *   <InputGroup>
+ *     <InputLeadingItems> … </InputLeadingItems>
+ *     <Input />
+ *     <InputTrailingItems> … </InputTrailingItems>
+ *   </InputGroup>
+ */
+export function InputGroup({children, ...inputProps}: InputProps) {
+  const [leadingWidth, setLeadingWidth] = useState<number>();
+  const [trailingWidth, setTrailingWidth] = useState<number>();
+
+  return (
+    <InputGroupContext.Provider
+      value={{inputProps, leadingWidth, setLeadingWidth, trailingWidth, setTrailingWidth}}
+    >
+      <InputGroupWrap disabled={inputProps.disabled}>{children}</InputGroupWrap>
+    </InputGroupContext.Provider>
+  );
+}
+
+export {InputProps};
+export const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
+  const {inputProps, leadingWidth, trailingWidth} = useContext(InputGroupContext);
+
+  return (
+    <StyledInput
+      ref={ref}
+      leadingWidth={leadingWidth}
+      trailingWidth={trailingWidth}
+      {...inputProps}
+      {...props}
+    />
+  );
+});
+
+/**
+ * Returns an array of media query matches, one for each breakpoint. We'll
+ * re-calculate the leading/trailing widths when any of these values changes.
+ */
+function useBreakpointMatches() {
```

Comment:
Is this necessary? Shouldn't the padding stay the same, regardless of the size of the viewport?

Reply:
There is a chance that one or more of the items are restyled to have different widths at different viewports, like:

```jsx
const StyledButton = styled(Button)`
  @media only screen and (max-width: ${p => p.theme.breakpoints.small}) {
      padding: …
   }
`
```

I don't think this is likely, but still added `useBreakpointMatches` just to be sure.

Reply:
Do you think we can remove this until it's clear we need it? I'm a bit concerned about the performance of all these event listeners. And aren't there other situations in which we might want to resize the padding? Like if a consumer of this component rendered one of the buttons conditionally - would that be handled?

Reply:
Sure thing! https://github.com/getsentry/sentry/pull/40270/commits/8f73916f07c4db36fb68b45066b8ac289424a7c5